File: Producto/back/src/main/java/PitagoraBackend/service/DashboardService.java
package PitagoraBackend.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import PitagoraBackend.repository.TicketsRepository;
import PitagoraBackend.repository.ObservacionesRepository;
import PitagoraBackend.repository.CategoriasRepository;
import PitagoraBackend.repository.ClientesRepository;
import PitagoraBackend.repository.ObrasRepository;
import PitagoraBackend.dto.DashboardStatsDTO;
import PitagoraBackend.dto.TopFallaDTO;
import PitagoraBackend.dto.ObraConIncidenciasDTO;
import PitagoraBackend.dto.ObraCostoDTO;
import PitagoraBackend.model.Categorias;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class DashboardService {
    
    @Autowired
    private TicketsRepository ticketsRepository;
    
    @Autowired
    private ObservacionesRepository observacionesRepository;
    
    @Autowired
    private CategoriasRepository categoriasRepository;

    @Autowired
    private ClientesRepository clientesRepository;

    @Autowired
    private ObrasRepository obrasRepository;
    
    /**
     * Obtiene las estadísticas generales del dashboard
     */
    public DashboardStatsDTO obtenerEstadisticas() {
        Long totalTickets = ticketsRepository.count();
        Long ticketsAbiertos = ticketsRepository.countTicketsAbiertos();
        Long observacionesAbiertas = observacionesRepository.countObservacionesAbiertas();
        Long observacionesTerminadas = observacionesRepository.countObservacionesTerminadas();
        Long observacionesAltaUrgencia = observacionesRepository.countObservacionesAltaUrgencia();
        Long clientesActivos = clientesRepository.countByEstado("Activo");
        Long obrasActivas = obrasRepository.countByEstadoObra("Activa");
        
        return new DashboardStatsDTO(
            totalTickets,
            ticketsAbiertos,
            observacionesAbiertas,
            observacionesTerminadas,
            observacionesAltaUrgencia,
            clientesActivos,
            obrasActivas
        );
    }
    
    /**
     * Obtiene el top 5 de fallas más reportadas
     */
    public List<TopFallaDTO> obtenerTopFallas() {
        return observacionesRepository.findTop5Fallas()
            .stream()
            .limit(5)
            .collect(Collectors.toList());
    }

    /**
     * Obtiene el top 5 de obras con mayor costo acumulado
     */
    public List<ObraCostoDTO> obtenerTopObrasPorCosto() {
        return observacionesRepository.findObrasPorCosto()
            .stream()
            .filter(o -> o.getMontoTotal() != null && o.getMontoTotal() > 0)
            .limit(5)
            .collect(Collectors.toList());
    }
    
    /**
     * Obtiene las obras asociadas a una categoría (drill-down)
     */
    public List<ObraConIncidenciasDTO> obtenerObrasPorCategoria(String nombreCategoria) {
        // Buscar la categoría por nombre
        List<Categorias> categorias = categoriasRepository.findByNombreCategoria(nombreCategoria);
        
        if (categorias.isEmpty()) {
            return List.of(); // Retorna lista vacía si no existe la categoría
        }
        
        Integer idCategoria = categorias.get(0).getIdCategoria();
        
        // Obtener obras asociadas a esta categoría
        return observacionesRepository.findObrasByCategoria(idCategoria);
    }
}

// Made with Bob

File: Producto/front/src/services/dashboardService.js
const API_URL = 'http://localhost:8080/api/dashboard';

const API_BASE = 'http://localhost:8080/api';

const contarObservacionesTerminadas = async () => {
  const response = await fetch(`${API_BASE}/observaciones`);
  if (!response.ok) return 0;
  const observaciones = await response.json();
  return observaciones.filter((o) =>
    (o.estadoObservacion || o.estado_observacion || '').toLowerCase() === 'terminado'
  ).length;
};

/**
 * Obtiene las estadísticas generales del dashboard
 * @returns {Promise<Object>} Objeto con totalTickets, ticketsAbiertos, observacionesAbiertas, observacionesTerminadas, etc.
 */
export const obtenerEstadisticas = async () => {
  try {
    const response = await fetch(`${API_URL}/stats`);
    if (!response.ok) {
      throw new Error('Error al obtener estadísticas');
    }
    const data = await response.json();

    let observacionesTerminadas = data.observacionesTerminadas;
    if (observacionesTerminadas === undefined || observacionesTerminadas === null) {
      observacionesTerminadas = await contarObservacionesTerminadas();
    }

    return {
      totalTickets: data.totalTickets ?? 0,
      ticketsAbiertos: data.ticketsAbiertos ?? 0,
      observacionesAbiertas: data.observacionesAbiertas ?? 0,
      observacionesTerminadas,
      observacionesAltaUrgencia: data.observacionesAltaUrgencia ?? 0,
      clientesActivos: data.clientesActivos ?? 0,
      obrasActivas: data.obrasActivas ?? 0,
    };
  } catch (error) {
    console.error('Error en obtenerEstadisticas:', error);
    throw error;
  }
};

/**
 * Obtiene el top 5 de fallas más reportadas
 * @returns {Promise<Array>} Array de objetos con nombreCategoria y cantidad
 */
export const obtenerTopFallas = async () => {
  try {
    const response = await fetch(`${API_URL}/top-fallas`);
    if (!response.ok) {
      throw new Error('Error al obtener top fallas');
    }
    return await response.json();
  } catch (error) {
    console.error('Error en obtenerTopFallas:', error);
    throw error;
  }
};

/**
 * Obtiene el top 5 de obras con mayor costo acumulado
 * @returns {Promise<Array>} Array de objetos con idObra, nombreObra y montoTotal
 */
export const obtenerTopObrasPorCosto = async () => {
  try {
    const response = await fetch(`${API_URL}/top-obras-costo`);
    if (!response.ok) {
      throw new Error('Error al obtener top obras por costo');
    }
    return await response.json();
  } catch (error) {
    console.error('Error en obtenerTopObrasPorCosto:', error);
    throw error;
  }
};

/**
 * Obtiene las obras asociadas a una categoría (drill-down)
 * @param {string} nombreCategoria - Nombre de la categoría
 * @returns {Promise<Array>} Array de objetos con idObra, nombreObra y cantidad
 */
export const obtenerObrasPorCategoria = async (nombreCategoria) => {
  try {
    const response = await fetch(`${API_URL}/obras-por-categoria?categoria=${encodeURIComponent(nombreCategoria)}`);
    if (!response.ok) {
      throw new Error('Error al obtener obras por categoría');
    }
    return await response.json();
  } catch (error) {
    console.error('Error en obtenerObrasPorCategoria:', error);
    throw error;
  }
};

// Made with Bob
